### PR TITLE
Dts writing

### PR DESCRIPTION
**Pull request by mhegazy**
Adding @RyanCavanaugh's documentation for writing declaration files.

### CODE REVIEW

File: pages/Writing Declaration Files.md [hunk 1]
```diff
@@ -1,268 +0,0 @@
-# Introduction
```
Comment (DanielRosenwasser) [inline pages/Writing Declaration Files.md]: There needs to be some sort of redirect here

Comment (mhegazy) [inline pages/Writing Declaration Files.md]: done.

File: pages/declaration files/By Example.md [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        title?: string;
+        color?: string;
+    }
+}
+```
+
+You can also created nested namespaces in one declaration:
+
+```ts
+declare namespace GreetingLib.Options {
```
Comment (DanielRosenwasser) [inline pages/declaration files/By Example.md]: I thought this was an anti-pattern.

Comment (mhegazy) [inline pages/declaration files/By Example.md]: why?

File: pages/declaration files/Introduction.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+The [By Example](./By Example.md) section shows many common API patterns and how to write declarations for each of them.
+This guide is aimed at the TypeScript novice who may not yet be familiar with every language construct in TypeScript.
+
+## Deep Dive
+
+For seasoned authors interested in the underlying mechanics of how declaration files work,
+  the [Deep Dive](./Deep Dive.md) section explains many advanced concepts in declaration writing,
+  and shows how to leverage these concepts to create cleaner and more intuitive declaration files.
+
+## `/templates`
```
Comment (DanielRosenwasser) [inline pages/declaration files/Introduction.md]: Is this even going to work on the website?

Comment (mhegazy) [inline pages/declaration files/Introduction.md]: i added a task to copy the templates.

Comment (mhegazy) [inline pages/declaration files/Introduction.md]: the folder link does not seem to work now that you mention it

Comment (mhegazy) [inline pages/declaration files/Introduction.md]: probably should remove it.

File: pages/declaration files/Deep Dive.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+  and a *value* `C` which refers to the constructor function of the class.
+Enum declarations behave similarly.
+
+### User Combinations
+
+Let's say we wrote a module file `foo.d.ts`:
+
+```ts
+export var SomeVar: { a: SomeType };
+export interface SomeVar {
```
Comment (devoto13) [inline pages/declaration files/Deep Dive.md]: This should be `export interface SomeType`, isn't it?

File: pages/declaration files/Deep Dive.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+
+### Built-in Combinations
+
+Astute readers will notice that, for example, `class` appeared in both the *type* and *value* lists.
+The declaration `class C { }` creates two things:
+  a *type* `C` which refers to the instance shape of the class,
+  and a *value* `C` which refers to the constructor function of the class.
+Enum declarations behave similarly.
+
+### User Combinations
```
Comment (devoto13) [inline pages/declaration files/Deep Dive.md]: Example from this section looks made-up to me. Maybe it worths replacing it with example with `class` and `namespace` (like one in [module-class.d.ts](https://github.com/Microsoft/TypeScript-Handbook/blob/dtsWriting/pages/declaration%20files/templates/module-class.d.ts))? It seems to be much more common and understandable pattern.

File: pages/Writing Declaration Files.md [hunk 6]
```diff
@@ -1,268 +1 @@
...
-#### Typing
-
-```ts
-// Note: 'void' return type is preferred here
-function addLater(x: number, y: number, callback: (sum: number) => void): void;
-```
-
-Please post a comment [here](https://github.com/Microsoft/TypeScript-Handbook/issues) if there's a pattern you'd like to see documented!
-We'll add to this as we can.
+> ## This page moved to [Publishing](./declaration files/Introduction.md)
```
Comment (DanielRosenwasser) [inline pages/Writing Declaration Files.md]: has been moved to

File: pages/declaration files/Library Structures.md [hunk 7]
```diff
@@ -0,0 +1,364 @@
+# Overview
+
+Broadly speaking, the way you *structure* your declaration file depends on how the library is consumed.
+There are many ways of offering a library for consumption in JavaScript, and you'll need to write your declaration file to match it.
+This guide covers how to identify common library patterns, and how to write declaration files which correspond to that pattern.
+
+Each type of major library structuring pattern has a corresponding file in the [`templates`](./templates) directory.
```
Comment (DanielRosenwasser) [inline pages/declaration files/Library Structures.md]: Templates will be broken

Comment (mhegazy) [inline pages/declaration files/Library Structures.md]: once this lands in master i will change the template links to point to the master version.

**Comment by DanielRosenwasser** [comment]
Seems decent :+1:

**Pull request merged**